go/storage/mkvs/db/pathbadger/pathbadger.go

```diff
@@ -0,0 +1,943 @@
+	if nonEmptyFinalizedRoots != nonEmptyVisitedRoots {
+		return fmt.Errorf("mkvs/pathbadger: not all finalized roots were visited (db corruption?)")
+	}
+
+	// Copy over any updated nodes for finalized roots with non-zero sequence numbers.
+	for rht, nodes := range notLoneNodes {
+		seqNo := finalizedSeqNos[rht]
+		if seqNo == 0 {
+			continue // Already in the right place.
+		}
+
+		for k := range nodes {
+			// Fetch pending node value. We will remove it later.
+			item, err := tx.Get(pendingNodeKeyFmt.Encode(version, rht, seqNo, []byte(k)))
+			if err != nil {
+				return fmt.Errorf("mkvs/pathbadger: failed to copy node: %w", err)
+			}
+			var value []byte
+			err = item.Value(func(data []byte) error {
+				value = append([]byte{}, data...) // Must copy to queue in batch.
+				return nil
+			})
+			if err != nil {
+				return fmt.Errorf("mkvs/pathbadger: failed to copy node: %w", err)
+			}
+
+			// Copy over to new location.
+			if err := batch.Set(finalizedNodeKeyFmt.Encode(rht, []byte(k)), value); err != nil {
+				return fmt.Errorf("mkvs/pathbadger: failed to copy node: %w", err)
+			}
+		}
+	}
+	// Remove any lone nodes.
+	for rht, nodes := range maybeLoneNodes {
+		for k := range nodes {
+			if _, isNotLone := notLoneNodes[rht][k]; isNotLone {
+				continue
+			}
+
+			if err := batch.Delete(finalizedNodeKeyFmt.Encode(rht, []byte(k))); err != nil {
+				return fmt.Errorf("mkvs/pathbadger: failed to delete lone node: %w", err)
+			}
+		}
+	}
+
+	// Remove all temporary nodes for non-zero sequence numbers.
+	pendingPrefix := pendingNodeKeyFmt.Encode(version)
+	pendingIt := tx.NewIterator(badger.IteratorOptions{Prefix: pendingPrefix})
+	defer pendingIt.Close()
+
+	for pendingIt.Rewind(); pendingIt.Valid(); pendingIt.Next() {
+		if err := batchMeta.Delete(pendingIt.Item().KeyCopy(nil)); err != nil {
+			return fmt.Errorf("mkvs/pathbadger: failed to delete pending node: %w", err)
+		}
+	}
+
+	pendingIt.Close()
+
+	// Remove any queued keys. Do this last as this action is irreversible in case of crash.
+	for _, key := range removeMetaKeys {
+		if err := batchMeta.Delete(key); err != nil {
+			return fmt.Errorf("mkvs/pathbadger: failed to delete key: %w", err)
+		}
+	}
+
+	// Commit batches.
+	if err := batch.Flush(); err != nil {
+		return err
+	}
+	if err := batchMeta.Flush(); err != nil {
+		return err
+	}
+
+	// Update last finalized version.
+	if err := d.meta.setLastFinalizedVersion(tx, version); err != nil {
+		return fmt.Errorf("mkvs/pathbadger: failed to set last finalized version: %w", err)
+	}
+
+	if err := tx.CommitAt(tsMetadata, nil); err != nil {
+		return fmt.Errorf("mkvs/pathbadger: failed to commit metadata: %w", err)
+	}
+
+	// Clean multipart metadata if there is any.
+	if d.multipartVersion != multipartVersionNone {
+		if err := d.cleanMultipartLocked(false); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Implements api.NodeDB.
+func (d *badgerNodeDB) Prune(version uint64) error {
```

Comment:
I would change this to work like this: prune last, hide, prune new.
- Add another field `lastPrunedVersion`.
- If `lastPrunedVersion == earliestVersion -2`, prune it, as it looks like the last pruning failed/was not completed.
- Set `earliestVersion = version+1`, which will hide `version`, but the pruning still needs to be done.
- Prune `version` and set `lastPrunedVersion = version`.

Reply:
Because if `d.meta.commit(tx)` at the end fails, we pruned but did not update the earliest version. So on restart, we will serve incomplete version. Therefore, I would always check before pruning the next one if the latest version was pruned.